src/WebJobs.Extensions.DurableTask/HttpApiHandler.cs

```diff
@@ -34,48 +35,65 @@ public HttpApiHandler(DurableTaskExtension config, TraceWriter traceWriter)
             {
-                // This is expected to include the auto-generated system key for this extension.
-                querySuffix += "&" + notificationUri.Query.TrimStart('?');
+                throw new ArgumentException($"Total timeout {timeout.TotalSeconds} should be bigger than retry timeout {retryInterval.TotalSeconds}");
             }
+            this.GetRetryValues(timeout, retryInterval, out var iterationCount, out var leftOverInterval);
+            JToken durableFunctionOutput = null;
 
-            string statusQueryGetUri = instancePrefix + "?" + querySuffix;
-            string sendEventPostUri = instancePrefix + "/" + RaiseEventOperation + "/{eventName}?" + querySuffix;
-            string terminatePostUri = instancePrefix + "/" + TerminateOperation + "?reason={text}&" + querySuffix;
-
-            HttpResponseMessage response = request.CreateResponse(
-                HttpStatusCode.Accepted,
-                new
+            var client = this.GetClient(request);
+            var status = await client.GetStatusAsync(instanceId);
+            for (var i = 0; i < iterationCount; i++)
+            {
+                await this.DelayExecution(i, iterationCount, retryInterval, leftOverInterval);
+                if (status != null && status.RuntimeStatus == OrchestrationRuntimeStatus.Completed)
                 {
-                    id = instanceId,
-                    statusQueryGetUri = statusQueryGetUri,
-                    sendEventPostUri = sendEventPostUri,
-                    terminatePostUri = terminatePostUri
-                });
+                    durableFunctionOutput = status.Output;
+                    break;
+                }
+                else
+                {
+                    if (status != null && (status.RuntimeStatus == OrchestrationRuntimeStatus.Canceled || status.RuntimeStatus == OrchestrationRuntimeStatus.Failed || status.RuntimeStatus == OrchestrationRuntimeStatus.Terminated))
+                    {
+                        break;
+                    }
+                    else
+                    {
+                        status = await client.GetStatusAsync(instanceId);
+                    }
+                }
+            }
 
-            // Implement the async HTTP 202 pattern.
-            response.Headers.Location = new Uri(statusQueryGetUri);
-            response.Headers.RetryAfter = new RetryConditionHeaderValue(TimeSpan.FromSeconds(10));
+            if (durableFunctionOutput == null)
+            {
+                if (status.RuntimeStatus == OrchestrationRuntimeStatus.Running)
+                {
+                    return this.CreateResponseMessage(request, instanceId, statusQueryGetUri, sendEventPostUri, terminatePostUri);
+                }
+                else
+                {
+                    return request.CreateErrorResponse(HttpStatusCode.InternalServerError, $"The durable function's runtime status is {status.RuntimeStatus}");
```

Comment:
I don't think this error message provides enough information. It might be a good idea to return the same JSON payload that `HandleGetStatusRequestAsync` returns, for consistency with the async polling pattern.

Reply:
Sure